### ISSUE TITLE
file transport without a git binary

### ISSUE THREAD

**Issue by petar** [go-git/go-git#838]
I discovered that go-git uses a local git binary when interacting with repositories on the local file system: https://github.com/go-git/go-git/blob/master/plumbing/transport/file/client.go#L19

Is there a way to remove the dependence on a git binary here? 

This dependency is introducing complications when trying to distribute portable software based on go-git. 

Thank you.

**Comment by lalyos** [comment]
Hi Petar,
This repo is mainly a library, but it has a cli also included: https://github.com/go-git/go-git/tree/master/cli/go-git

What I did was, instead of installing git by the os package manager, I just used the "go-git" binary, and I was able to git push/pull via ssh protocol.

Either you can get the binary via: `go install github.com/go-git/go-git/cli/go-git@latest`
Or you can make some wrapper function which will call:
- git-receive: https://github.com/go-git/go-git/blob/master/cli/go-git/receive_pack.go#L24
- and https://github.com/go-git/go-git/blob/master/cli/go-git/upload_pack.go#L25

**Comment by petar** [comment]
@lalyos Thanks for your note.

I investigated option (3). It appears that the go-git cli simply uses `server.DefaultServer` under the hood (https://github.com/go-git/go-git/blob/master/plumbing/transport/file/server.go#L16 and https://github.com/go-git/go-git/blob/master/plumbing/transport/file/server.go#L34)

This makes me wonder. Can one just override the default file client to use `server.DefaultServer`, like so:

```go
import (
	"github.com/go-git/go-git/v5/plumbing/transport/client"
	"github.com/go-git/go-git/v5/plumbing/transport/server"
)

func init() {
	client.Protocols["file"] = server.DefaultServer
}
```

**Comment by petar** [comment]
When I try using the above code, I get an "object not found" error during the push operation.

I tried also using:

```go
client.InstallProtocol("file", server.NewClient(server.DefaultLoader))
```

But I am getting the same error. I am still not entirely sure why.

**Comment by pjbgf** [comment]
> I discovered that go-git uses a local git binary when interacting with repositories on the local file system

@petar that's correct, the [compatibility file](https://github.com/go-git/go-git/blob/master/COMPATIBILITY.md?plain=1#L212) calls that out. However, this is a dependency that we should get rid of if/when we can.

> But I am getting the same error. I am still not entirely sure why.

I don't know all the ins-and-outs of that part of the code, so it would need some investigating. Would you be keen on implementing this?

If we were to decommission the underlying `git` binary requirement, it would be nice to also provide a way to set the rootfs for the "remote" - as `DefaultLoader` uses `/`.

**Comment by pjbgf** [comment]
@petar I did a quick test and it seems to work:

```go
func main() {
	url := "file:///go-git-bare"
	localDir, err := os.MkdirTemp("", "go-git")
	CheckIfError(err)

	targetRoot := "/home/coder/git/"

	Info("%q in target root %q to local dir %q", url, targetRoot, localDir)

	loader := server.NewFilesystemLoader(osfs.New(targetRoot, osfs.WithBoundOS()))
	client.InstallProtocol("file", server.NewClient(loader))

	Info("git clone %s %s", url, localDir)

	r, err := git.PlainClone(localDir, false, &git.CloneOptions{
		URL: url,
	})
	CheckIfError(err)

	ref, err := r.Head()
	CheckIfError(err)

	commit, err := r.CommitObject(ref.Hash())
	CheckIfError(err)

	fmt.Println(commit)
}
```

Before the running below, a bare version of go-git was placed at `/home/coder/git/go-git-bare`:
```sh
$ git
zsh: command not found: git

$ go run _examples/clone/main.go
"file:///go-git-bare" in target root "/home/coder/git/" to local dir "/tmp/go-git3751941336"
git clone file:///go-git-bare /tmp/go-git3751941336
/home/coder/git/go-git-bare/config
commit 8c1e3e2eb458677af127fc286a43ad262c3cca1e
...
```

But I am not sure about the UX of setting the target root the way I did.

**Comment by onee-only** [comment]
Note that the `v6-transport` branch includes a refactored version of file transport, which doesn't use git binary.
Once it's merged, I think we can close this issue.

**Comment by pjbgf** [comment]
@onee-only good point. I will go ahead and close this, as that is already merged and tracked for `v6`, and we won't be back-porting that to `v5`.
